### PR TITLE
RFC1004: Add `try_transaction` and `retry`

### PR DESCRIPTION
**Pull request by tailhook**
Like in python PR edgedb/edgedb-python#146 this lacks documentations and also more robustness can be added later. But this provides correct future-proof API.

### CODE REVIEW

File: src/pool.ts [hunk 1]
```diff
@@ -273,12 +274,12 @@ class PoolConnectionHolder {
   }
 }
 
-const holderAttr = Symbol.for("holder");
-const detach = Symbol.for("detach");
-export const getHolder = Symbol.for("getHolder");
-const connectionAttr = Symbol.for("connection");
-export const unwrapConnection = Symbol.for("unwrap");
-const isDetached = Symbol.for("isDetached");
+const holderAttr = Symbol("holder");
```
Comment (1st1) [inline src/pool.ts:277]: Good catch!

File: src/ifaces.ts [hunk 2]
```diff
@@ -75,10 +83,16 @@ interface Modifiable {
 export type Executor = ReadOnlyExecutor & Modifiable;
 
 export interface Connection extends Executor {
+  [BORROW]?: BorrowReason;
```
Comment (1st1) [inline src/ifaces.ts:86]:
```suggestion
  [BORROWED_FOR]?: BorrowReason;
```

File: src/client.ts [hunk 3]
```diff
@@ -145,40 +229,59 @@ class StandaloneConnection implements Connection {
...
   async query(query: string, args?: QueryArgs): Promise<Set> {
+    this._check_borrow();
     let connection = this._connection;
     if (!connection || connection.isClosed()) {
       connection = await this._reconnect();
     }
     return await connection.query(query, args);
   }
   async queryJSON(query: string, args?: QueryArgs): Promise<string> {
+    this._check_borrow();
```
Comment (1st1) [inline src/client.ts:248]: JS JITs don't inline functions that raise errors. So I'd guard with `if (this[BORROW]) {`.

File: src/client.ts [hunk 4]
```diff
@@ -145,40 +229,59 @@ class StandaloneConnection implements Connection {
     return this._isClosed;
   }
   async execute(query: string): Promise<void> {
```
Comment (1st1) [inline src/client.ts:231]: Please add newlines between methods

**Review by 1st1 — COMMENTED** [review]
Looks good, let's continue fixing few of the remaining nits

File: src/client.ts [hunk 5]
```diff
@@ -111,16 +123,88 @@ class StandaloneConnection implements Connection {
...
+          continue;
+        }
+        throw err;
+      }
+      // TODO(tailhook) sort out errors on commit, early network errors
+      // and some other errors could be retried
+      // NOTE: we can't retry on all the same errors as we don't know if
+      // commit is succeeded before the database have received it or after
+      // it have been done but network is dropped before we were able
+      // to receive a response
```
[the remark this answers is not in the document]
Comment (tailhook) [reply src/client.ts:201]: "before we were were to receive a response". I.e. before client received the CommandComplete for the `COMMIT`.

File: src/client.ts [hunk 6]
```diff
@@ -69,6 +74,12 @@ enum TransactionStatus {
   TRANS_UNKNOWN = 4, // cannot determine status
 }
 
+const DEFAULT_MAX_ITERATIONS = 3;
```
[the remark this answers is not in the document]
Comment (tailhook) [reply src/client.ts:77]: I think we agreed no to do it this round.

File: src/client.ts (hunk 3 above)
Comment (tailhook) [reply src/client.ts:247]: Done

File: src/client.ts (hunk 4 above)
Comment (tailhook) [reply src/client.ts:249]: I've added whenever found. But honestly I don't understand why we have a prettifier if it doesn't handle such a basic thing for us.

File: src/ifaces.ts (hunk 2 above)
Comment (tailhook) [reply src/ifaces.ts]: Done

**Review by 1st1 — APPROVED** [review]

**Pull request merged**